### ISSUE TITLE
[Bug]: Seeded model produces varying results

### ISSUE THREAD

**Issue by Kallinteris-Andreas** [DLR-RM/stable-baselines3#1480]
### 🐛 Bug

I seeded my model, and yet it does produce deterministic results.
Do I need to seed the env (that shouldn't be the case since the model will be resetting the env)?
Maybe the `EvalCallback` is not deterministic (it does have a seed argument).
The warning also do not appear to have any useful info.

Or am I missing something obvious?

Note: the training was done on the CPU not CUDA

Thanks!

### To Reproduce

```python
import os
import gymnasium as gym

from stable_baselines3 import TD3
from stable_baselines3.common.callbacks import EvalCallback
from stable_baselines3.common.env_util import make_vec_env

env_id = "InvertedDoublePendulum-v4"

eval_log_dir = "./results/"
os.makedirs(eval_log_dir, exist_ok=True)

env = gym.make(env_id)
eval_env = gym.make(env_id)

eval_callback = EvalCallback(eval_env, best_model_save_path=eval_log_dir,
                              log_path=eval_log_dir, eval_freq=500,
                              n_eval_episodes=5, deterministic=True,
                              render=False)

model = TD3("MlpPolicy", env, seed=0)
model.learn(5000, callback=eval_callback)
```

### Relevant log output / Error message

```shell
$ py test.py
/home/master-andreas/.local/lib/python3.10/site-packages/gymnasium/envs/registration.py:693: UserWarning: WARN: Overriding environment GymV26Environment-v0 already in registry.
  logger.warn(f"Overriding environment {new_spec.id} already in registry.")
/home/master-andreas/.local/lib/python3.10/site-packages/torch/cuda/__init__.py:83: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:109.)
  return torch._C._cuda_getDeviceCount() > 0
/home/master-andreas/.local/lib/python3.10/site-packages/stable_baselines3/common/evaluation.py:67: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(
Eval num_timesteps=500, episode_reward=33.60 +/- 3.35
Episode length: 3.80 +/- 0.40
New best mean reward!
/home/master-andreas/.local/lib/python3.10/site-packages/stable_baselines3/common/evaluation.py:67: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(
Eval num_timesteps=1000, episode_reward=31.94 +/- 4.13
Episode length: 3.60 +/- 0.49
Eval num_timesteps=1500, episode_reward=33.56 +/- 3.34
Episode length: 3.80 +/- 0.40
Eval num_timesteps=2000, episode_reward=30.19 +/- 4.09
Episode length: 3.40 +/- 0.49
Eval num_timesteps=2500, episode_reward=116.74 +/- 33.67
Episode length: 12.60 +/- 3.61
New best mean reward!
Eval num_timesteps=3000, episode_reward=107.29 +/- 13.59
Episode length: 11.60 +/- 1.50
Eval num_timesteps=3500, episode_reward=120.47 +/- 25.07
Episode length: 13.00 +/- 2.68
New best mean reward!
Eval num_timesteps=4000, episode_reward=125.97 +/- 13.93
Episode length: 13.60 +/- 1.50
New best mean reward!
Eval num_timesteps=4500, episode_reward=191.43 +/- 52.69
Episode length: 20.60 +/- 5.64
New best mean reward!
Eval num_timesteps=5000, episode_reward=248.88 +/- 61.17
Episode length: 26.80 +/- 6.52
New best mean reward!



$ py test.py
/home/master-andreas/.local/lib/python3.10/site-packages/gymnasium/envs/registration.py:693: UserWarning: WARN: Overriding environment GymV26Environment-v0 already in registry.
  logger.warn(f"Overriding environment {new_spec.id} already in registry.")
/home/master-andreas/.local/lib/python3.10/site-packages/torch/cuda/__init__.py:83: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:109.)
  return torch._C._cuda_getDeviceCount() > 0
/home/master-andreas/.local/lib/python3.10/site-packages/stable_baselines3/common/evaluation.py:67: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(
Eval num_timesteps=500, episode_reward=31.90 +/- 4.06
Episode length: 3.60 +/- 0.49
New best mean reward!
/home/master-andreas/.local/lib/python3.10/site-packages/stable_baselines3/common/evaluation.py:67: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(
Eval num_timesteps=1000, episode_reward=31.87 +/- 4.10
Episode length: 3.60 +/- 0.49
Eval num_timesteps=1500, episode_reward=30.25 +/- 4.16
Episode length: 3.40 +/- 0.49
Eval num_timesteps=2000, episode_reward=33.56 +/- 3.32
Episode length: 3.80 +/- 0.40
New best mean reward!
Eval num_timesteps=2500, episode_reward=118.51 +/- 30.82
Episode length: 12.80 +/- 3.31
New best mean reward!
Eval num_timesteps=3000, episode_reward=107.35 +/- 22.66
Episode length: 11.60 +/- 2.42
Eval num_timesteps=3500, episode_reward=150.20 +/- 56.17
Episode length: 16.20 +/- 6.05
New best mean reward!
Eval num_timesteps=4000, episode_reward=125.85 +/- 24.14
Episode length: 13.60 +/- 2.58
Eval num_timesteps=4500, episode_reward=174.38 +/- 56.83
Episode length: 18.80 +/- 6.05
New best mean reward!
Eval num_timesteps=5000, episode_reward=264.13 +/- 58.68
Episode length: 28.40 +/- 6.28
New best mean reward!
```

### System Info
```sh
$ pip list | grep stabl
stable-baselines3             2.0.0a5

$ python --version          
Python 3.10.10

$ pip list | grep torch
torch                         1.12.1

$ pip list | grep gym  
gymnasium                     0.28.1
```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
I tried but couldn't reproduce the issue using the RL Zoo (master version):
`python -m rl_zoo3.train --algo td3 --env InvertedDoublePendulum-v4 -n 5000 --eval-freq 500 --seed 0 -P --verbose 0 -c tmp.yml`

where `tmp.yml` is:
```yaml
InvertedDoublePendulum-v4:
  policy: "MlpPolicy"
  n_timesteps: 5000
  train_freq: 2
  gradient_steps: 1
  learning_starts: 100
```
(I changed slightly the hyperparameters for faster training but this should not change the result of whether it is deterministic or not)

I think the issue is that you don't seed your eval env.

**Comment by Kallinteris-Andreas** [comment]
Is there a way to make such that on every `eval _allback` the first evaluation run is reset with `seed=64`, the second with `seed=65`... 

Motivation: I want to benchmark minor differences in the `Hopper-v4` and `Hopper-v5` and need all `eval_callback’s to evaluate the same domains

**Comment by araffin** [comment]
> Is there a way to seed the first reset of each EvalCallback event?

You need to have a custom callback.

> n: I want to benchmark minor differences in the Hopper-v4 and Hopper-v5

Seeding the first episode won't solve your problem as Hopper can have a variable number of steps.
What you need to do is to use a quantitative number of evaluation episodes (and you can use multiple envs at the same time to make things faster).
This is easily done using the RL Zoo (see below for the command).
You can also compare the results to the Open RL SB3 benchmark: https://github.com/openrlbenchmark/openrlbenchmark/issues/7
W&B page: https://wandb.ai/openrlbenchmark/sb3

```
python -m rl_zoo3.train --algo <ALGO> --env <ENV> --eval-episodes 20 --n-eval-envs 5 --track
```

**Comment by araffin** [comment]
PS: if you want to run things even faster, you could use SBX: https://rl-baselines3-zoo.readthedocs.io/en/master/guide/sbx.html

**Comment by Kallinteris-Andreas** [comment]
1) Gymnasium environments do not increment the reset seed during stepping only during resetting so the number of steps the environment runs for does not matter
(at least that is the case for MuJoCo environments)

2) I have been avoiding using RL zoo, because it has a few too many abstractions for my liking  

3) I do not want to run SBX

4) [proposal] add a `seed` argument to `evaluate_policy` (and by extension `EvalCallback`) where only the first env `reset()` would be seeded with the `seed`

**Comment by araffin** [comment]
> Gymnasium environments do not increment the reset seed during stepping only during resetting so the number of steps the environment runs for does not matter

I meant you are trying to compare policy A and B.
For that you are seeding the envs using `seed=0` for instance.

The first evaluation episode will start in the same state, but the episode may end at different steps for A and B.
This means that the second episode doesn't start in the same state for A and B, even though you seeded the first episode.

> I have been avoiding using RL zoo, because it has a few too many abstractions for my liking.

It is a black box indeed but it is battle-tested (so it prevents errors like the first one you had) and perfect for benchmarking and having reproducible runs.

> (like for example what seed does the training_env and eval_env use)

The exact seed doesn't matter, what matters is that the two experiments start in the same state (algo and envs seeded the same between two experiments), and that's what the `--seed` argument allows.

> would it be possible to add a seed: int argument to

Again, this wouldn't be useful (see reply to your first point), and seeding the env before passing it to evaluate policy would have the exact same effect.

**Comment by Kallinteris-Andreas** [comment]
1) episode length does not impact the `env.reset()` (this might be specific to `mujoco` envs, I am not sure)
```python
import gymnasium

env = gymnasium.make('Hopper-v4')

action=[0.2, -0.3, 0.1] # some random action
env.reset(seed=0)
env.step(action)
env.step(action)
env.step(action)
obs0, _ = env.reset()  # Reset after 3 steps

env.reset(seed=0)
env.step(action)
env.step(action)
env.step(action)
env.step(action)
env.step(action)
env.step(action)
obs1, _ = env.reset()  # Reset after 5 steps

assert (obs0 == obs1).all()  # Does not fail
```

2)Here is the policy evaluator (from a different project), this is the behavior, I want to achieve
Each policy evaluation run is run on the same starting states.
```python
# Runs policy for X episodes and returns return reward
# A fixed seed is used for the eval environment
def eval_policy(env_name: str, seed: int = 256, eval_episodes: int = 10) -> float:
    eval_env = gymnasium.make(env_name)

    total_return = 0
    for i in range(eval_episodes):
        state = eval_env.reset(seed=seed + i)[0]    # <-- Each evalution run evaluates on the same starting states
        terminated, truncated = 0, 0
        while not (terminated or truncated):
            action = model.query_actor(torch.tensor(state, dtype=torch.float32, device=TORCH_DEVICE), add_noise=False)
            state, reward, terminated, truncated, _ = eval_env.step(action.tolist())
            total_return += reward

    return total_return / eval_episodes
```




3) to clarify 3 "cases" of seeding the eval_env 
- [a]
```python
env_eval = ...
for every_freq:
    env_eval.reset(seed=my_seed)
    evaluate_policy(episodes=10)
```

- [b]
```python
    env_eval = ...
for every_freq:
    evaluate_policy(episodes=10, seed_of_first_episode=my_seed)
```

- [c]
```python
env_eval = ...
env_eval.reset(seed=my_seed)
for every_freq:
    evaluate_policy(episodes=10)
![image 8ELC41](https://user-images.githubusercontent.com/30759571/236176676-daeb2e41-0208-4249-87b4-6efdbcf06e8c.png)
![image](https://user-images.githubusercontent.com/30759571/236176682-80ef9dfb-585d-4f5c-ba81-78f96b197936.png)

```
- [d]
```python
env_eval = ...
for every_freq:
    evaluate_policy(episodes=10, seeds=list_of_10_ints)
```


Case [c] has the problem of each `evaluate_policy` call evaluating different starting states.
Cases [a] [b] [d] do not have that problem, and any of would be acceptable  (with preferences going [a]<[b]<[d])
The current implementation of `stable_baselines3.common.callbacks.EvalCallback` only allows for [c].
If `EvalCallback` was modified with a `seed` argument then case [a] could be achieved.
If `stable_baselines3.common.evaluation.evaluate_policy` was modified then [b] and [d] could be achieved.

**Comment by araffin** [comment]
> does not impact the env.reset() (this might be specific to mujoco envs, I am not sure)

Interesting... yes probably specific to MuJoCo.

closing as the original question was answered.

> was modified then [b] and [d] could be achieved.

As I wrote, for your use case, best is to write a custom callback so you have full control over the evaluation.
(Although I really think having enough eval episodes + runs (different seeds) should be sufficient for benchmark purposes)

**Comment by Kallinteris-Andreas** [comment]
> closing as the original question was answered.

Ok.

@araffin I will modify `EvalCallback` for my needs, would you like a PR with the `seed` argument, or should I keep it to my self?

**Comment by araffin** [comment]
> or should I keep it to my self?

For now, keep it for yourself.
